### ISSUE TITLE
Re-consider "no trailing slash" URLs over "trailing slash" URLs

### ISSUE THREAD

**Issue by shea256** [pyeve/eve#118]
Options:
- No trailing slash: `/users` and `/users/1`
- Trailing slash: `/users/` and `/users/1/`

After doing a quick google search, I found a bunch of resources that use the former:
- http://www.slideshare.net/calamitas/restful-best-practices
- http://madhatted.com/2013/3/19/suggested-rest-api-practices
- http://www.vinaysahni.com/best-practices-for-a-pragmatic-restful-api
- https://blog.apigee.com/detail/restful_api_design_nouns_are_good_verbs_are_bad
- http://www.restapitutorial.com/lessons/restfulresourcenaming.html
- http://stackoverflow.com/questions/12208117/restful-like-crud-operation-url-pattern-for-nested-model
- http://architects.dzone.com/news/common-rest-design-pattern

This one says "either":
- http://googlewebmastercentral.blogspot.com/2010/04/to-slash-or-not-to-slash.html

AngularJS doesn't even support trailing slashes:
- http://docs.angularjs.org/api/ngResource.$resource

And here is an interesting reference about the original paper on REST:
- http://programmers.stackexchange.com/questions/186959/trailing-slash-in-restful-api
- http://en.wikipedia.org/wiki/Representational_state_transfer#RESTful_web_APIs

Pretty much all the top APIs when you google "api" (Twitter, Instagram, Foursquare, Tumblr) don't use a trailing slash:
- https://dev.twitter.com/docs/api/1.1
- http://instagram.com/developer/endpoints/comments/
- https://developer.foursquare.com/docs/users/lists
- http://www.tumblr.com/docs/en/api/v2

**Comment by snuggs** [comment]
**TL;DR;**
> In fact this practice is so pervasive, that for the vast majority of users, a resource URI with or without a trailing slash is treated as a synonym. They are considered two URIs that point to the same resource, using either one is fine. However this understanding is not quite correct...

https://cdivilly.wordpress.com/2014/03/11/why-trailing-slashes-on-uris-are-important/

@shea256 with all due respect this is horrible information, incorrect, an anti-pattern, nor is it the "industry standard". However I do fully respect the fact this is commonplace. Doesn't mean it's right. The only correct way to satisfy your result is with a `301` redirect of either or. There are no alternatives (aside from `<link rel=canonical ...>`.

@GDmac actually stated the "industry standard". I'll always believe the spec first. (smart) Men lie (and don't even know it), women like, specs do not. `/foo` and `/foo/` are INDEED two resources.

Technically speaking I can `PUT /dogs` and 

1. `PUT` CAN have side effects to other resources `/dogs` != `/dogs/`
2. Many people (even super smart developers) don't know about the [`Content-Location`](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Location) header.

Just because we do it doesn't mean we are right. And just because others around us are doing it doesn't make it more right. That's not social validation that's an echo chamber.

There are two types of resources. Canonical resources and "subresources". `/foo/bar` where `bar` is a **subresource** of `/foo/` and `/foo/bar/` where `bar/` is (still) a subresource of `/foo/` but now a bonafied resource in and of itself. 

#FunFact: Place create a `foo/` directory and a `foo/bar/` directory. Place an `index.html` file in each folder. Place a `cat.gif` image in the `foo/bar/` path. Put `<img alt='' src=cat.gif>` in **both** `index.html` files. The web will show you exactly why these are "two different unique resources".

[Further validation](https://github.com/fastlane/ci/issues/15#issuecomment-373597905)

Trust me I used to think the same way about the time you wrote your statement. But facts don't seem to care about my opinion FWIW.
